src/lib.rs

```diff
@@ -150,7 +150,7 @@ impl<K, R: SafeRng> EnvelopeCipher<K, R> {
     }
 }
 
-impl<K: KeyProvider, R: SafeRng> EnvelopeCipher<K, R> {
+impl<K: KeyProvider<U16>, R: SafeRng> EnvelopeCipher<K, R> {
```

Comment:
Could you make this generic and choose Aes128 or Aes256 depending on the key size? Might save extra duplication but unsure of your plans here

Reply:
I tried doing this. I have some problems:
1. The key size `S` is part of the trait generic param, and not available in code, so I can't do a comparison check of `if S == U16` or something similar like a `match`.
2. key size `S` is of arbitrary length (I found `U0` to `U10000000000000000000`), so for `U16`, `U32`, or any arbitrary `S`, I can't really create a `cipher` and assign it to a common variable `cipher`. There isn't a common trait that I can use to put it in a `Box<dyn Trait>` easily.
3. Suppose I can map `U16` to `aes128`, and `U32` to `aes256`, what do I do with the other arbitrary key sizes `S`? I feel that I'm not clear in this area.

Reply:
I'm starting to think that, maybe instead of arbitrary key size `S` here, we create our own enum KeySize of just `16` and `32`, that way we can check it in code correctly. What do you think?

Reply:
@ragibkl the key size should not be arbitrary - I agree that it should be limited to 16 or 32 bytes. In which case that can be expressed as an enum with two different values.

Reply:
I'll try with an enum in a separate branch then. This PR should probably be back to draft.